daml-script/export/src/main/scala/com/daml/script/export/TreeUtils.scala

```diff
@@ -440,6 +440,25 @@ object TreeUtils {
     )
   }
 
+  def moduleRefs(
+      acs: Iterable[CreatedEvent],
+      trees: Iterable[TransactionTree],
+  ): Set[String] = {
+    val fromAcs = acs.foldMap(ev => valueRefs(Sum.Record(ev.getCreateArguments)))
+    val fromTrees = trees.foldMap(treeRefs(_))
+    (fromAcs ++ fromTrees)
+      .map(_.moduleName)
+      .filterNot(isInternalModule)
+  }
+
+  private def isInternalModule(moduleName: String): Boolean = {
+    // Keep in sync with isInternal in compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor.hs
+    moduleName.startsWith("DA.Internal.") ||
+    moduleName.startsWith("GHC.") ||
+    Set("Control.Exception.Base", "Data.String", "LibraryModules", "DA.Types", "DA.Time.Types")
```

Comment:
Are you sure this works correctly? `DA.Time.Types` defines `RelTime` which is not exported by `Prelude`. As far as I can tell if I have a reference to `RelTime` this will now just result in an import error.

For `DA.Internal.Template` you are lucky because the `Archive` type is reexported from `Prelude`.

I haven’t checked if there are other types to which this applies in addition to `RelTime` but I suspect so.

The easiest fix would be to allow imports of these modules. While that sounds perhaps a bit odd, I think you can actually justify this:
1. The module names are already part of the public API if you’re working with the ledger API.
2. Any attempts at hiding constructors also don’t apply when working over the ledger API so I think providing an option to construct things without going through smart constructors directly in Daml is only reasonable.

wdyt? @sofiafaro-da I’d also appreciate your input here

Reply:
Another possibility is to rewrite these imports:

- DA.Time.Types -> DA.Time
- GHC.Stack.Types -> DA.Stack
- DA.Internal.Exception -> DA.Exception
- DA.Internal.Record -> DA.Record

Everything else seems to be exposed in Prelude, is hidden, or is special (e.g. Erased and PromotedText).

Making these four modules non-internal would be difficult without making all the other modules non-internal. But I'm also ok with us dropping the concept of "internal module" (that can't be imported) in the preprocessor, since that's not a thing in LF. The module name and lack of official docs should be enough to dissuade people to not import them directly unless they know what they're doing and are fine with being burned.

Reply:
Rewriting imports works only partially, For Daml script exports we will still fallover when we need to use the constructor for `RelTime` for example which is currently hidden from `DA.Time`.

Reply:
Got it. I'm happy for us to remove the internal module distinction altogether. (In any case, internal modules should not be considered a "stable" part of our standard library. Except for the ones that are in stable packages.)

Reply:
I think for the purposes of this discussion, stable packages are the only ones that matter. We will only get references to serializable types from the ledger and those are all in stable packages. Fully agree that having those in internal modules doesn’t make sense.

Reply:
Yes, that's a good point. In case of `RelTime` we could also issue `DA.Time.convertMicrosecondsToRelTime` to work around the issue. But, trying to keep up with all the necessary rewrites sounds brittle.
All in favor of simply allowing these imports.
I'm happy to look into this.